##### Title = "Mr"
##### Name = "Muhammad Sami-Ur-Rahman"
##### Email = "Samiurr87@gmail.com"

## DAY-17
### Topic 
 
                    

# Data Wrangling 

- Handling Missing Value 
- Data formatting 
- Data Normalization
    - Scaling 
    - Centralizing 
- Data Bining 
    - For Groups of Data
- Making Dummies of Categorical data
    - Categorical to Numerical 

In [277]:
# STEP-1 IMPORT LAIBRARIES 
import numpy as np 
import pandas as pd
import seaborn as sns 
import matplotlib.pyplot as plt 

In [278]:
# STEP-2 Load titanic Data set from sns
boat = sns.load_dataset("titanic")
boatextra = boat
boatextra2 = boat
boat.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [186]:
# Simple Operation (Math Operator) along the columns 

# boat["age"]+10 


### Dealing and Handling Missing Value 
- Missing Values are N/A, NaN or 0 or Blank Cell 
- When Data is not in row and in any perameter 

>

> Steps:

1. Koshish karen dobara data collect kar len ya dekh len agar kahin ghalti hy

2.0

2. Missing value wala variable (column) hi nikal den agar data per effect nahi hta ya simple row or data entry remove kar den

3. Replace the missing values:

1. How?

1. Average value of entire variable or similar data point

2. frequencey or MODE replacement

3. Replace based on other functions (Data sampler knows that)

4. ML algortihm can also be used

5. Leave it like that


Why Do we need to do that?.
- its better because no data will be loss 
- Our Result will be more accurate

In [187]:
# STEP-3 fIND MISSING VALUE WITH SUM
boat.isnull().sum()
# AS wE CAN SEE "AGE" HAS 177 MISSING VALUE AND DECL HAS 688 

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [188]:
# STEP-5    LETS TAKE "DROP.NA" FUNCTION 

boat.shape


(891, 15)

In [189]:
boat.dropna(subset=['deck'],axis=0,inplace=True) # this code will specifically remove the Na Values from deck columns. 
# Inplace = True modifies the data frame 
boat.shape

(203, 15)

In [190]:
boat.isnull().sum()

survived        0
pclass          0
sex             0
age            19
sibsp           0
parch           0
fare            0
embarked        2
class           0
who             0
adult_male      0
deck            0
embark_town     2
alive           0
alone           0
dtype: int64

In [191]:
# STEP-6 Now we will drop Na Values from Whole data and store this new clean dataset into "boat1"
newboat = boat.dropna()
 # Remove from Whole Data frame 
newboat.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           0
embark_town    0
alive          0
alone          0
dtype: int64

In [192]:
newboat.shape

(182, 15)

# Replacing Missing Values with the 
# average of that column 

In [193]:
# First checking Missing Value
boatextra.isnull().sum()

survived        0
pclass          0
sex             0
age            19
sibsp           0
parch           0
fare            0
embarked        2
class           0
who             0
adult_male      0
deck            0
embark_town     2
alive           0
alone           0
dtype: int64

In [194]:
# Finding an average (mean)
mean = boatextra['age'].mean()
mean

35.77945652173913

In [195]:
#Replacing Nan with mean of the data data will be update as well.
boatextra['age'] = boatextra['age'].replace(np.nan, mean)

In [196]:
boatextra.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
deck           0
embark_town    2
alive          0
alone          0
dtype: int64

In [197]:
percent_missing = boatextra.isnull().sum() * 100 / len(boat1)
percent_missing

survived       0.000000
pclass         0.000000
sex            0.000000
age            0.000000
sibsp          0.000000
parch          0.000000
fare           0.000000
embarked       0.224467
class          0.000000
who            0.000000
adult_male     0.000000
deck           0.000000
embark_town    0.224467
alive          0.000000
alone          0.000000
dtype: float64

 Here we Can See there are "688" Null Values Out of Total 891 Values. Data is missing more than 77%. Soo we will delete this column.

In [198]:
boatextra.drop(['deck'], inplace=True, axis=1)

In [199]:
boatextra.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

In [202]:
# For Replacing "embark_town" Nan values with mode because its Categorical variable 
mode = boatextra['embark_town'].mode()
mode


0    Southampton
dtype: object

In [203]:
boatextra.embark_town

1        Cherbourg
3      Southampton
6      Southampton
10     Southampton
11     Southampton
          ...     
871    Southampton
872    Southampton
879      Cherbourg
887    Southampton
889      Cherbourg
Name: embark_town, Length: 203, dtype: object

Fillna() Methods for Categorical/Class Veriable
            For Replacing "embark_town" Nan values

In [204]:
# fillna()
boatextra["embark_town"]=boatextra['embark_town'].fillna("Southampton")
boatextra

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
6           0       1    male  54.0      0      0  51.8625        S  First   
10          1       3  female   4.0      1      1  16.7000        S  Third   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
871         1       1  female  47.0      1      1  52.5542        S  First   
872         0       1    male  33.0      0      0   5.0000        S  First   
879         1       1  female  56.0      0      1  83.1583        C  First   
887         1       1  female  19.0      0      0  30.0000        S  First   
889         1       1    male  26.0      0      0  30.0000        C  First   

       who  adult_male  embark_town alive  alone  
1    woman       False    Cherbourg   yes  False  
3    woman       False  Southampton   yes  False  
6      man        True  Southampton    no   True  
10   child       False  Southampton   yes  False  
11   woman       False  Southampton   yes   True  
..     ...         ...          ...   ...    ...  
871  woman       False  Southampton   yes  False  
872    man        True  Southampton    no   True  
879  woman       False    Cherbourg   yes  False  
887  woman       False  Southampton   yes   True  
889    man        True    Cherbourg   yes   True  

[203 rows x 14 columns]

In [205]:
boatextra.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

In [206]:
boatextra.embarked.mode()

0    S
dtype: object

In [207]:
boatextra["embarked"] = boatextra["embarked"].fillna('S')
boatextra.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

# Data Formatting 
- To cahange Data on one/common standard (Data ko ik format per lana)

- Ensures data is consitent and understandabale
    - Easy to gather
    - Easy to work with
        - Faisalabad (FSD)
        - Lahore (LHR)
        - Islamabad (ISB)
        - Karachi (KCH)
        - Peshawar (PEW)
        - Convert g to kg or similar uni for all
        - one standard unit in each column
            - ft != cm

In [208]:
boatextra2.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
embark_town      object
alive            object
alone              bool
dtype: object

In [209]:
# Changing Data type from one set of data to other (Type Casting)
boatextra2['survived'] = boatextra2['survived'].astype('float64')
boatextra2.dtypes

survived        float64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
embark_town      object
alive            object
alone              bool
dtype: object

In [210]:
# Changing Data type from one set of data to other 
boatextra2['survived'] = boatextra2['survived'].astype('int64')
boatextra2.dtypes

survived          int64
pclass            int64
sex              object
age             float64
sibsp             int64
parch             int64
fare            float64
embarked         object
class          category
who              object
adult_male         bool
embark_town      object
alive            object
alone              bool
dtype: object

In [211]:
boatextra2['age'] = boatextra2['age']*365
boatextra2

survived  pclass     sex      age  sibsp  parch     fare embarked  class  \
1           1       1  female  13870.0      1      0  71.2833        C  First   
3           1       1  female  12775.0      1      0  53.1000        S  First   
6           0       1    male  19710.0      0      0  51.8625        S  First   
10          1       3  female   1460.0      1      1  16.7000        S  Third   
11          1       1  female  21170.0      0      0  26.5500        S  First   
..        ...     ...     ...      ...    ...    ...      ...      ...    ...   
871         1       1  female  17155.0      1      1  52.5542        S  First   
872         0       1    male  12045.0      0      0   5.0000        S  First   
879         1       1  female  20440.0      0      1  83.1583        C  First   
887         1       1  female   6935.0      0      0  30.0000        S  First   
889         1       1    male   9490.0      0      0  30.0000        C  First   

       who  adult_male  embark_town alive  alone  
1    woman       False    Cherbourg   yes  False  
3    woman       False  Southampton   yes  False  
6      man        True  Southampton    no   True  
10   child       False  Southampton   yes  False  
11   woman       False  Southampton   yes   True  
..     ...         ...          ...   ...    ...  
871  woman       False  Southampton   yes  False  
872    man        True  Southampton    no   True  
879  woman       False    Cherbourg   yes  False  
887  woman       False  Southampton   yes   True  
889    man        True    Cherbourg   yes   True  

[203 rows x 14 columns]

In [212]:
boatextra2['age'] = boatextra2['age'].astype('int64')
boatextra2.rename(columns={"age":"age in days"}, inplace=True)
boatextra2

survived  pclass     sex  age in days  sibsp  parch     fare embarked  \
1           1       1  female        13870      1      0  71.2833        C   
3           1       1  female        12775      1      0  53.1000        S   
6           0       1    male        19710      0      0  51.8625        S   
10          1       3  female         1460      1      1  16.7000        S   
11          1       1  female        21170      0      0  26.5500        S   
..        ...     ...     ...          ...    ...    ...      ...      ...   
871         1       1  female        17155      1      1  52.5542        S   
872         0       1    male        12045      0      0   5.0000        S   
879         1       1  female        20440      0      1  83.1583        C   
887         1       1  female         6935      0      0  30.0000        S   
889         1       1    male         9490      0      0  30.0000        C   

     class    who  adult_male  embark_town alive  alone  
1    First  woman       False    Cherbourg   yes  False  
3    First  woman       False  Southampton   yes  False  
6    First    man        True  Southampton    no   True  
10   Third  child       False  Southampton   yes  False  
11   First  woman       False  Southampton   yes   True  
..     ...    ...         ...          ...   ...    ...  
871  First  woman       False  Southampton   yes  False  
872  First    man        True  Southampton    no   True  
879  First  woman       False    Cherbourg   yes  False  
887  First  woman       False  Southampton   yes   True  
889  First    man        True    Cherbourg   yes   True  

[203 rows x 14 columns]

# Data Normalization 
Normalize data is Bell sahpe curve data
- Uniform the data 
- they have same impact 
- Also for computational reasons
- Data should be in one range , therfor we can compare them more better

In [221]:
ks1=boatextra2[['age in days','fare']]
ks1.head()

age in days     fare
1         13870  71.2833
3         12775  53.1000
6         19710  51.8625
10         1460  16.7000
11        21170  26.5500

- The above data is really in wide range and we need to normalize and hard to compare

- Normalization change the values to the range of 0-to-1 (now both variable has similar influence on our models)

Method of Normalization

1. Simple feature scaling 
    - x(new)= x(old)/x(max)
2. Min-Max Methods
3. Z-score (standard Score) -3 to +3
4. Log Transformation

In [225]:
# Simple Feature Scaling / Normaliza "Fare" and "age in days"
ks1['fare']= ks1['fare']/ks1['fare'].max() 
ks1['age in days']= ks1['age in days']/ks1['age in days'].max()

ks1.head()

C:\Users\M6205~1.TAN\AppData\Local\Temp/ipykernel_160/3869361384.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ks1['fare']= ks1['fare']/ks1['fare'].max()


age in days      fare
1        0.4750  0.139136
3        0.4375  0.103644
6        0.6750  0.101229
10       0.0500  0.032596
11       0.7250  0.051822

In [228]:
# Min - Max Methods 
ks1['fare']= (ks1['fare']-ks1['fare'].min())/(ks1['fare'].max()-ks1['fare'].min())
ks1.head()


C:\Users\M6205~1.TAN\AppData\Local\Temp/ipykernel_160/3179490476.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ks1['fare']= (ks1['fare']-ks1['fare'].min())/(ks1['fare'].max()-ks1['fare'].min())


age in days      fare
1        0.4750  0.139136
3        0.4375  0.103644
6        0.6750  0.101229
10       0.0500  0.032596
11       0.7250  0.051822

In [304]:
# Loading Data set
ks2=sns.load_dataset('titanic')
ks2

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [305]:
# Z-Score (Standard score)
ks2['fare']= (ks2['fare']-ks2['fare'].mean())/(ks2['fare'].std())
ks2.head()

survived  pclass     sex   age  sibsp  parch      fare embarked  class  \
0         0       3    male  22.0      1      0 -0.502163        S  Third   
1         1       1  female  38.0      1      0  0.786404        C  First   
2         1       3  female  26.0      0      0 -0.488580        S  Third   
3         1       1  female  35.0      1      0  0.420494        S  First   
4         0       3    male  35.0      0      0 -0.486064        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [306]:
# Log Transformation
ks2['fare'] = np.log(ks2['fare'])
ks2.head()

C:\Users\M.Tanzeel\anaconda3\lib\site-packages\pandas\core\arraylike.py:364: RuntimeWarning: invalid value encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


survived  pclass     sex   age  sibsp  parch      fare embarked  class  \
0         0       3    male  22.0      1      0       NaN        S  Third   
1         1       1  female  38.0      1      0 -0.240285        C  First   
2         1       3  female  26.0      0      0       NaN        S  Third   
3         1       1  female  35.0      1      0 -0.866325        S  First   
4         0       3    male  35.0      0      0       NaN        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

Binning
- Grouping of values into smaller number of values (bins)
- Convert numeric into categories (jawan, Bachay, Boorhay) or 1-16, 17-30 etc
- To have better understanding of groups
    - low vs mid vs high price

In [313]:
bins = np.linspace (min (ks2['age']), max(ks2['age']), 15000) 
agegroups = ["Bachay", "Jawan", "Boorhay"]
ks2['age'] = pd.cut(ks2['age'], bins=bins, labels=agegroups, include_lowest=True)
ks2['age']

ValueError: Bin labels must be one fewer than the number of bin edges

In [296]:
# bins = np.linspace (min (ks2['age']), max(ks2['age']), 15000) 
# age groups = ["Bachay", "Jawan", "Boorhay"]
# ks2['age'] = pd.cut (ks2['age' ], bins, labels-age_groups, include_lowest-True)
# ks2.['age']

#### Converting Ctaegories Into Dummies Value
- easy to use for computation
- Malre female (0,1)

In [297]:
ks2.sex

0        male
1      female
2      female
3      female
4        male
        ...  
886      male
887    female
888    female
889      male
890      male
Name: sex, Length: 891, dtype: object

In [301]:
dummy = pd.get_dummies(ks2['sex'])
# ks2 = pd.concat([ks2['sex'],dummy])
df=pd.concat([ks2['sex'],dummy],axis=1)
df

sex  female  male
0      male       0     1
1    female       1     0
2    female       1     0
3    female       1     0
4      male       0     1
..      ...     ...   ...
886    male       0     1
887  female       1     0
888  female       1     0
889    male       0     1
890    male       0     1

[891 rows x 3 columns]